internal/gps/result.go

```diff
@@ -42,37 +45,61 @@ type solution struct {
 	solv Solver
 }
 
-// WriteDepTree takes a basedir and a Lock, and exports all the projects
-// listed in the lock to the appropriate target location within the basedir.
+// WriteDepTree takes a baseDir and a Lock, and exports all the projects
+// listed in the lock to the appropriate target location within baseDir.
 //
-// If the goal is to populate a vendor directory, basedir should be the absolute
+// If the goal is to populate a vendor directory, baseDir should be the absolute
 // path to that vendor directory, not its parent (a project root, typically).
 //
-// It requires a SourceManager to do the work, and takes a flag indicating
-// whether or not to strip vendor directories contained in the exported
-// dependencies.
-func WriteDepTree(basedir string, l Lock, sm SourceManager, sv bool) error {
+// It requires a SourceManager to do the work, and takes a PruneOptions
+// indicating the pruning options required for the exported dependencies.
+func WriteDepTree(baseDir string, l Lock, sm SourceManager, prune PruneOptions, logger *log.Logger) error {
+	if baseDir == "" {
+		return errors.New("must provide a non-empty baseDir")
+	}
 	if l == nil {
-		return fmt.Errorf("must provide non-nil Lock to WriteDepTree")
+		return errors.New("must provide a non-nil Lock to WriteDepTree")
+	}
+	if sm == nil {
+		return errors.New("must provide a non-nil SourceManager to WriteDepTree")
 	}
 
-	err := os.MkdirAll(basedir, 0777)
-	if err != nil {
+	if err := os.MkdirAll(baseDir, 0777); err != nil {
 		return err
 	}
 
-	// TODO(sdboyer) parallelize
+	var wg sync.WaitGroup
+	errCh := make(chan error, len(l.Projects()))
+
 	for _, p := range l.Projects() {
-		to := filepath.FromSlash(filepath.Join(basedir, string(p.Ident().ProjectRoot)))
+		wg.Add(1)
+		go func(p LockedProject) {
+			to := filepath.FromSlash(filepath.Join(baseDir, string(p.Ident().ProjectRoot)))
 
-		err = sm.ExportProject(p.Ident(), p.Version(), to)
-		if err != nil {
-			removeAll(basedir)
-			return fmt.Errorf("error while exporting %s: %s", p.Ident().ProjectRoot, err)
-		}
-		if sv {
-			filepath.Walk(to, stripVendor)
+			if err := sm.ExportProject(p.Ident(), p.Version(), to); err != nil {
+				removeAll(to)
+				errCh <- errors.Wrapf(err, "failed to export %s", p.Ident().ProjectRoot)
+			}
+
+			wg.Done()
+		}(p)
+	}
+
+	wg.Wait()
+
+	if len(errCh) > 0 {
+		logger.Println("Failed to write dep tree. The following errors occurred:")
+		for err := range errCh {
```

Comment:
Since channel is not closed, this will be blocked? Channel should be closed after `wg.Wait()` and return below should return an error since closed channel yields nil.